### PR TITLE
Automatically archive Pulses when last subscription is deleted

### PR DESCRIPTION
**Pull request by camsaul**
This is a little more complicated than it needs to be because Toucan 1 doesn't have a `post-delete` hook, and since you can't easily  write logic for when a CRUD operation happens for a child object (e.g. write code to do something for a `Pulse` when one of its `PulseChannel`s is deleted). You can do this sort of stuff in Toucan 2 btw.

The solution I went with here is to have to `PulseChannelRecipient` call a `will-delete-recipient` function in the `models.pulse-channel` namespace when it's about to be deleted, which decides whether to delete the channel as well; and to have `PulseChannel` call a `will-delete-channel` function in the `models.pulse` namespace when it's about to be deleted, which decides whether to archive the Pulse as well. By doing things this was I was able to keep Pulse- and PulseChannel-related logic in their respective namespaces.

### CODE REVIEW

File: test/metabase/cmd/rotate_encryption_key_test.clj [hunk 1]
```diff
@@ -13,11 +13,14 @@
...
             [metabase.test :as mt]
             [metabase.test.data.interface :as tx]
+            [metabase.test.fixtures :as fixtures]
             [metabase.util.encryption :as encrypt]
             [metabase.util.encryption-test :as eu]
             [toucan.db :as db]
             [toucan.models :as models]))
 
+(use-fixtures :once (fixtures/initialize :db))
+
```
Comment (camsaul) [inline test/metabase/cmd/rotate_encryption_key_test.clj:23]: unrelated to this PR but this should fix occasional test failures of `rotate-encryption-key!-test` (hopefully)

File: src/metabase/models/pulse_channel.clj [hunk 2]
```diff
@@ -171,18 +194,18 @@
...
+                 [:= :schedule_hour hour]]
+                [:and [:= :schedule_type "weekly"]
+                 [:= :schedule_hour hour]
+                 [:= :schedule_day weekday]]
+                [:and [:= :schedule_type "monthly"]
+                 [:= :schedule_hour hour]
+                 [:= :schedule_frame schedule-frame]
+                 [:or [:= :schedule_day weekday]
+                  ;; this is here specifically to allow for cases where day doesn't have to match
+                  [:= :schedule_day monthly-schedule-day-or-nil]]]]]})))
```
Comment (camsaul) [inline src/metabase/models/pulse_channel.clj:208]: unrelated minor indentation fix

File: src/metabase/models/pulse.clj [hunk 3]
```diff
@@ -346,13 +360,12 @@
...
-  have keys like `id`, `include_csv`, and `include_xls`.
+  "Update the PulseCards for a given `notification-or-id`. `card-refs` should be a definitive collection of *all* Cards
+  for the Notification in the desired order. They should have keys like `id`, `include_csv`, and `include_xls`.
 
-   *  If a Card ID in `card-refs` has no corresponding existing `PulseCard` object, one will be created.
-   *  If an existing `PulseCard` has no corresponding ID in CARD-IDs, it will be deleted.
-   *  All cards will be updated with a `position` according to their place in the collection of `card-ids`"
+  *  If a Card ID in `card-refs` has no corresponding existing `PulseCard` object, one will be created.
+  *  If an existing `PulseCard` has no corresponding ID in CARD-IDs, it will be deleted.
+  *  All cards will be updated with a `position` according to their place in the collection of `card-ids`"
```
Comment (camsaul) [inline src/metabase/models/pulse.clj:368]: unrelated indentation fix

**Review by dpsutton — APPROVED** [review]
